source/iterator/cdc.go

```diff
@@ -0,0 +1,411 @@
+						// shouldn't happen cause VEventType_FIELD always comes before VEventType_ROW.
+						sdk.Logger(ctx).Warn().Msgf("i.fields is nil, skipping the row")
+
+						continue
+					}
+
+					rowEvent = event
+
+				default:
+				}
+			}
+		}
+	}
+}
+
+// processRowEvent makes rows from the event.RowEvent.RowChanges trusted,
+// constructs the resulting slice containing all needed sqltypes.Values,
+// transforms it to a sdk.Record and sends the record to a c.records channel.
+func (c *cdc) processRowEvent(ctx context.Context, event *binlogdata.VEvent) error {
+	for _, change := range event.RowEvent.RowChanges {
+		var (
+			valuesBefore []sqltypes.Value
+			valuesAfter  []sqltypes.Value
+			operation    = sdk.OperationCreate
+		)
+
+		switch after, before := change.After, change.Before; {
+		case after != nil && before != nil:
+			operation = sdk.OperationUpdate
+			valuesBefore = sqltypes.MakeRowTrusted(c.fields, before)
+			valuesAfter = sqltypes.MakeRowTrusted(c.fields, after)
+
+		case before != nil:
+			operation = sdk.OperationDelete
+			valuesAfter = sqltypes.MakeRowTrusted(c.fields, before)
+
+		default:
+			valuesAfter = sqltypes.MakeRowTrusted(c.fields, after)
+		}
+
+		record, err := c.transformRowsToRecord(ctx, c.fields, valuesBefore, valuesAfter, operation)
+		if err != nil {
+			return fmt.Errorf("transform rows to record: %w", err)
+		}
+
+		c.records <- record
+	}
+
+	return nil
+}
+
+// transformRowsToRecord transforms after and before of type []sqltypes.Values to a sdk.Record,
+// based on provided fields and operation.
+func (c *cdc) transformRowsToRecord(
+	ctx context.Context, fields []*query.Field, before, after []sqltypes.Value, operation sdk.Operation,
+) (sdk.Record, error) {
+	var (
+		transformedRowBeforeBytes []byte
+		transformedRowAfterBytes  []byte
+		orderingColumnValue       any
+		key                       sdk.StructuredData
+		err                       error
+	)
+
+	if len(before) > 0 {
+		_, _, transformedRowBeforeBytes, err = c.transformValuesToNative(ctx, before)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("transform values to native: %w", err)
+		}
+	}
+
+	if len(after) > 0 {
+		key, orderingColumnValue, transformedRowAfterBytes, err = c.transformValuesToNative(ctx, after)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("transform values to native: %w", err)
+		}
+
+		// set this in order to avoid the 'same position' error,
+		// as the event can have multiple rows under one gtid.
+		c.position.LastProcessedElementValue = orderingColumnValue
+	}
+
+	sdkPosition, err := c.position.MarshalSDKPosition()
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal position to sdk position: %w", err)
+	}
+
+	metadata := make(sdk.Metadata)
+	metadata.SetCreatedAt(time.Now())
+	metadata[metadataKeyTable] = c.table
+
+	switch operation {
+	case sdk.OperationCreate:
+		return sdk.Util.Source.NewRecordCreate(sdkPosition, metadata, key, sdk.RawData(transformedRowAfterBytes)), nil
```

Comment:
It feels like it would be more natural if this was sdk.StructuredData.

Reply:
Updated.